src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs

```diff
@@ -13,10 +14,10 @@ namespace System.Text.Json
     [DebuggerDisplay("{DebuggerDisplay,nq}")]
     public readonly partial struct JsonElement
     {
-        internal readonly object _parent;
+        internal readonly object? _parent;
         private readonly int _idx;
 
-        internal JsonElement(JsonDocument parent, int idx)
+        internal JsonElement(JsonDocument? parent, int idx)
         {
             // parent is usually not null, but the Current property
             // on the enumerators (when initialized as `default`) can
```

Comment:
I think one of the nullable guidances was that if a struct field is only null for `default` that it should be treated non-nullable.  Only the TokenType property and CheckValidInstance should bother actually checking it for null.

Reply:
> I think one of the nullable guidances was that if a struct field is only null for default that it should be treated non-nullable

Not necessarily.  It depends on expected usage.  We start from a position of "if it can be null, it should be nullable", and then back off from that if any such usage would be deemed invalid and it really should be nullable.  For example, CancellationToken contains a CancellationTokenSource; it's absolutely fine and even expected to use default(CancellationToken), which is just CancellationToken.None, and its contained CancellationTokenSource absolutely must be nullable.  In contrast, the System.Reflection.InterfaceMap struct is never expected to be used with default(InterfaceMap), only as the return type of the reflection methods that hand one back, so while we started with its four public reference field types as being nullable, we decided to make them non-nullable, to represent expected usage.

Reply:
Fair enough... I'd say that this one is "expected non-null", since everything other than TokenType calls CheckValidInstance, and CheckValidInstance throws when this field is null.

Reply:
So make it non null then? that would save lots of bangs.

Reply:
The benefit of marking `_parent` as nullable is that folks who add APIs or make changes to the code will be forced to address the null warning (particularly if they forget the `CheckValidInstance`).

If we wanted to avoid the bangs, we could annotate the `CheckValidInstance` to help there. 

My original comment here was specific to the internal `JsonElement` ctor which forces `_parent` to be null **outside** of the `JsonElement elem = default` case when someone uses the `default` enumerator and calls `Current` on it. Maybe we should throw in those specific cases and leave the internal ctor as non-nullable `JsonDocument`.

That is somewhat separate from the `JsonElement elem = default` discussion. Given we have tests for default `JsonElement` case, I think leaving `_parent` it as nullable is OK. Though, I don't feel too strongly about it.

Reply:
> ... particularly if they forget the `CheckValidInstance`

Forgetting CheckValidInstance is just wrong.  It means that InvalidOperationException wouldn't be thrown on a `default` value, which they should add a test for.  If they fail to add it, they'll get a NullReferenceException, and then the fix is to call CheckValidInstance and have the exception be thrown.

Describing the field as non-null seems to match what I want the caller code to look like (call CheckValidInstance and then know that you're happy), without caring how the nullability feature is implemented.

Reply:
> Forgetting CheckValidInstance is just wrong.  It means that InvalidOperationException wouldn't be thrown on a `default` value, which they should add a test for.  If they fail to add it, they'll get a NullReferenceException, and then the fix is to call CheckValidInstance and have the exception be thrown.

Sure. The point with annotating it with nullability is that the contributor will catch the issue at build time rather than at test/run time. They may not add a test (and requires PR reviewers to point it out).

> Describing the field as non-null seems to match what I want the caller code to look like (call CheckValidInstance and then know that you're happy), without caring how the nullability feature is implemented.

Sure, that's fine by me. I was mainly pointing out what we lose out by not marking it as nullable, which to me seemed like the main value prop of using the nullability feature to begin with.

Reply:
Update from offline discussion (and what @bartonjs stated above): the `_parent` field doesn't need to be null and similarly, this internal ctor doesn't need to have the `parent` parameter be marked as nullable.

The only case where `_parent` can be null is if `JsonElement elem = default`, but all its instance methods throw in that case.

Reply:
Will update, do you want to keep asserts I added when `_parent` used without calling CheckValidInstance()?

Reply:
I only see one assert in `JsonElement.cs`.

Yes, let's keep that one.


What other asserts are there?

Reply:
There is 2 for each in `ArrayEnumerator and ObjectEnumerator` https://github.com/dotnet/runtime/pull/528/files#diff-f2d23490ae729f8360d702787ce036b9R27 and https://github.com/dotnet/runtime/pull/528/files#diff-3e144ca59d3df140dfddbc9b9652148eR29

Reply:
Looks good. Marking as resolved.